File: November2021/LargestDivisibleSubset.java
class Solution {
	public List<Integer> largestDivisibleSubset(int[] nums) {
		int[] dp = new int[nums.length];
		return constructLDS(nums, dp, getLDSSize(nums, dp));
	}

	private int getLDSSize(int[] nums, int[] dp) {
		Arrays.sort(nums);
		Arrays.fill(dp, 1);
		int ldsSize = 1;

		for (int i = 1; i < nums.length; i++)
			for (int j = 0; j < i; j++)
				if (nums[i] % nums[j] == 0) {
					dp[i] = Math.max(dp[i], dp[j] + 1);
					ldsSize = Math.max(ldsSize, dp[i]);
				}

		return ldsSize;
	}

	private List<Integer> constructLDS(int[] nums, int[] dp, int ldsSize) {
		int prev = -1;
		LinkedList<Integer> lds = new LinkedList<Integer>();

		for (int i = dp.length - 1; i >= 0; i--)
			if (dp[i] == ldsSize && (prev == -1 || prev % nums[i] == 0)) {
				lds.addFirst(nums[i]);
				ldsSize--;
				prev = nums[i];
			}

		return lds;
	}
}

File: November2021/C++/Largest_Divisible_Subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        vector<int> dp(nums.size(),1);
        int max_count=countLDS(dp,nums);
        // cout<<max_count;
        return getlds(nums,dp,max_count);
    }
    
    vector<int> getlds(vector<int>& nums,vector<int> dp,int& max_count)
    {
        vector<int> res;
        int prev=-1;
        for(int i=dp.size()-1;i>=0;i--)
        {
            // cout<<dp[i]<<endl;
            if(dp[i]==max_count && (prev==-1 || (prev%nums[i])==0))
            {
                
                res.push_back(nums[i]);
                prev=nums[i];
                --max_count;
            }
        }
        
        reverse(begin(res),end(res));
        return res;
    }
    int countLDS(vector<int>& dp,vector<int>& nums)
    {
        sort(begin(nums),end(nums));

        for(int i=1;i<nums.size();i++)
        {
            for(int j=0;j<i;j++)
            {
                if(nums[i]%nums[j]==0)
                    dp[i]=max(dp[i],dp[j]+1);
            }
        }
        
        return *max_element(begin(dp),end(dp));
    }
};